lib/Language/Souffle/Interpreted.hs

```diff
@@ -0,0 +1,215 @@
+
+{-# OPTIONS_GHC -Wno-redundant-constraints #-}
+{-# LANGUAGE TypeFamilies, DerivingVia, InstanceSigs #-}
+module Language.Souffle.Interpreted
+  ( Program(..)
+  , Fact(..)
+  , Marshal(..)
+  , Handle
+  , ContainsFact
+  , MonadSouffle(..)
+  , SouffleM
+  , runSouffle
+  , cleanup
+  ) where
+
+import Prelude hiding (init)
+
+import Control.DeepSeq (deepseq)
+import Control.Monad.State.Strict
+import Data.IORef
+import Data.Foldable (traverse_)
+import Data.List hiding (init)
+import Data.Maybe (fromMaybe)
+import Data.Proxy
+import Data.Word
+import Language.Souffle.Class
+import Language.Souffle.Marshal
+import System.Directory
+import System.Environment
+import System.FilePath
+import System.IO.Temp
+import System.Process
+import Text.Printf
+
+
+newtype SouffleM a = SouffleM { runSouffle :: IO a }
+  deriving
+    ( Functor
+    , Applicative
+    , Monad
+    , MonadFail
+    , MonadIO
+    ) via IO
```

Comment:
As out of curiosity, why do you use the deriving via?

Reply:
I used it in the compiled version as well and wanted it to be more consistent. It feels a little more descriptive than `GeneralizedNewtypeDeriving` and also more powerful (you can also do `DeriveAnyClass` and "stock" deriving with it also). 
I think the main drawback of `DerivingVia` is that it's only supported on newer GHCs.

See https://ryanglscott.github.io/talk-slides/deriving-via-plclub-slides.pdf for some better formulated reasons 😅.

Reply:
I was lucky to work with Baldur when he came up with the idea of the deriving via. But I didn't see too much codes using it yet. Probably because I don't really look into lot of haskell libraries nowadays.